containers: add kubernetes images scanning

simon3z opened the pull request.

simon3z commented:
@miq-bot add_label providers/containers
@miq-bot add_label enhancement
@miq-bot add_label wip

chessbyte commented on app/models/kubernetes_image_scan.rb:
```diff
@@ -0,0 +1,139 @@
+# Copyright 2015 Red Hat, Inc.
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
```
Why is this whole license header here?  Why Lesser GPL, when our project is Apache licensed?

blomquisg commented on app/models/kubernetes_image_scan.rb:
```diff
@@ -0,0 +1,139 @@
+
+class KubernetesImageScan < Job
+  PROVIDER_CLASS = ManageIQ::Providers::Kubernetes::ContainerManager
+  FLEECING_IMAGE = "docker.io/fsimonce/docker-fleece:latest"
```
I'm guessing that this line makes this a WIP pr.

simon3z commented:
@miq-bot remove_label wip

**NOTE:** this is currently blocked on #4217 and #4167 but other than that it's already possible to start reviewing to get it merged ASAP.

@matthewd @tenderlove can you check the ruby style (and anything else you may find)
@roliveri can you check that I am using the smart state analysis framework in the correct way
@gtanzillo can you check that the I am using the Job infrastructure in the best way possible at this time

cc @chessbyte 

Thanks.

simon3z commented on app/models/kubernetes_image_scan.rb:
```diff
@@ -0,0 +1,264 @@
+      end
+
+      _log.info("pod #{pod_full_name} is not available")
+      sleep(POD_POLL_INTERVAL)
```
@gtanzillo according to your suggestions, do you think it would be better to wait using the queue (even possible?)?
I actually have mixed feelings about it (again if possible). We could try to discuss it together.

roliveri commented on app/models/kubernetes_image_scan.rb:
```diff
@@ -0,0 +1,264 @@
+    )
+
+    image.perform_metadata_scan(ost)
+    image.perform_metadata_sync(ost)
```
Why are you bypassing image.scan_metadata/image.sync_metadata and the server_smart_proxy?

simon3z replied:
@roliveri because the code in `scan_metadata` and `.sync_metadata` are very much vm-specific and I am not using an host/smart_proxy for the analysis.

matthewd commented on app/models/kubernetes_image_scan.rb:
```diff
@@ -0,0 +1,264 @@
+require 'securerandom'
+
+class KubernetesImageScan < Job
+  PROVIDER_CLASS = ManageIQ::Providers::Kubernetes::ContainerManager
```
This class should be namespaced underneath `ManageIQ::Providers::Kubernetes::ContainerManager`

matthewd commented on app/models/container_image.rb:
```diff
@@ -1,10 +1,19 @@
+require 'MiqContainer/MiqContainer'
+
 class ContainerImage < ActiveRecord::Base
```
These changes sound entirely Not Okay; `ContainerImage` is provider independent, and shouldn't contain the word "docker"... and it certainly shouldn't contain the K word.

blomquisg replied:
:/ yeah, this will start to perpetuate the problems we have with VMWare logic being spread all over the place in Vms and Hosts.

Probably makes more sense to create a `ManageIQ::Providers::Kubernetes::ContainerManager::ContainerImage` to contain much of this logic.

simon3z replied:
@matthewd @blomquisg I removed the K words.
Anyway since now we just have docker images I would rather spare the effort of creating an STI (db migration, etc) and allow `docker_id` to live here.

roliveri commented on app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb:
```diff
@@ -0,0 +1,271 @@
+    )
+
+    image.perform_metadata_scan(ost)
+    image.perform_metadata_sync(ost)
```
@simon3z in what way are `scan_metadata` and `sync_metadata` VM-specific?
I know you're not using the `server_smart_proxy` (not host/smart_proxy), that's why I asked why you were bypassing it.

simon3z commented:
> @simon3z in what way are scan_metadata and sync_metadata VM-specific?
> I know you're not using the server_smart_proxy (not host/smart_proxy), that's why I asked why you were bypassing it.

@roliveri `scan_metadata` and `sync_metadata` are VM-specific because they assume you have [a smart proxy `host`](https://github.com/ManageIQ/manageiq/blob/2a59f4c/app/models/mixins/scanning_mixin.rb#L162) on which you'll `queue_call`. Tracing back where `host` comes from, you'll see that it is generated from `agent_class` and `agent_id` [here](https://github.com/ManageIQ/manageiq/blob/2a59f4c/app/models/vm_scan.rb#L85) and then continuing back you'll hit `job_proxy_dispatcher` setting the `agent_class` and `agent_id` [here](https://github.com/ManageIQ/manageiq/blob/2a59f4c/app/models/job_proxy_dispatcher.rb#L124-L127), [here](https://github.com/ManageIQ/manageiq/blob/2a59f4c/app/models/job_proxy_dispatcher.rb#L115), then following the `proxy` [here](https://github.com/ManageIQ/manageiq/blob/2a59f4c/app/models/job_proxy_dispatcher.rb#L82), [here](https://github.com/ManageIQ/manageiq/blob/2a59f4c/app/models/job_proxy_dispatcher.rb#L67), [here](https://github.com/ManageIQ/manageiq/blob/2a59f4c/app/models/job_proxy_dispatcher.rb#L66) you'll get to `get_eligible_proxies_for_job` [here](https://github.com/ManageIQ/manageiq/blob/2a59f4c/app/models/job_proxy_dispatcher.rb#L275) which requires `proxies4job` [here](https://github.com/ManageIQ/manageiq/blob/2a59f4c/app/models/job_proxy_dispatcher.rb#L298)

As you see there are many dependencies on `JobProxyDispatcher` (which is very much VM-oriented) if we use `scan_metadata` and `sync_metadata`.

The idea was to avoid `JobProxyDispatcher` since it requires a large refactor at this point and it's not clear if will be maintained in the future. As you said:

> While no special dispatch is needed for Container SSA, the JobProxyDispatcher is still required to initiate the job.

Actually it is possible to initiate the job also without the `JobProxyDispatcher` (as in this PR).

> There has also been talk of removing the JobProxyDispatcher completely, so we’ll have to determine the best course to take.

Given the uncertain fate of `JobProxyDispatcher` and the fact that there's no need to have a dispatcher for containers, I'd stay away from that part.

simon3z commented:
Ping @roliveri @chessbyte

chessbyte commented:
@roliveri @simon3z suggest you guys organize a meeting to discuss approaches and see if you can come to mutual agreement

roliveri commented:
@simon3z So, there are basically 2 issues here: 1. the `host` setting, and 2. `proxies4job` issue. Neither of these make the code in question VM-specific - in that they can apply to things that aren't VMs - and each can easily be accommodated by your code.

Actually, the setting of `host` - as I mentioned in my email on 8/28 - will solve both issues. All you need to do is pass `MiqServer.my_server` as the value for `host`. This will use the `server_smart_proxy` to perform the scan. Doing so will also eliminate the need for `get_eligible_proxies_for_job`, `proxies4job` and the `JobProxyDispatcher`.

You need to go through the common smart proxy and queuing mechanism, so your scans will use the same worker configurations as all the other scan types - configuration global to all SmartProxy scans.

simon3z commented:
@roliveri check if everything is fine now. If that's ok then I'll try to fix/polish some TODOs and we'd be ready for merge. Thanks. cc @chessbyte

roliveri commented:
@simon3z looks good to me.

On a side note, if you could identify things you needed to implement/re-implement because they weren't accessible to non VMs - like `queue_signal` - I'd like to factor those out for use in the future (post merge of this PR).

simon3z commented:
Ping review/merge @roliveri @blomquisg @chessbyte

Fryguy commented on app/models/manageiq/providers/kubernetes/container_manager/scanning/job.rb:
```diff
@@ -0,0 +1,292 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
```
Please remove the license header.  It is already covered by the repo LICENSE files.  Besides, it's wrong anyway since we are dual licensed at this time

Fryguy commented on gems/pending/MiqContainerGroup/MiqContainerGroup.rb:
```diff
@@ -0,0 +1,36 @@
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
```
Same comment on license header.

Fryguy commented on gems/pending/MiqContainerGroup/MiqContainerGroup.rb:
```diff
@@ -0,0 +1,36 @@
+
+require 'fs/MiqFS/modules/WebDAV'
+
+class MiqContainerGroup
```
This feels like a bad name/location, especially under gems/pending.  I feel like this more belongs in the fs tree somehow.

Fryguy replied:
Is this more of a ContainerGroupFs?

roliveri replied:
I don't think so. I think MiqContainerGroup corresponds to MiqVm - only it doesn't need to do much. The FS here is the WebDAV FS module.

Fryguy commented on db/migrate/20150904132351_add_scan_columns_to_container_images.rb:
```diff
@@ -0,0 +1,6 @@
+class AddScanColumnsToContainerImages < ActiveRecord::Migration
+  def change
+    add_column :container_images, :last_sync_on, :datetime
+    add_column :container_images, :last_scan_attempt_on, :datetime
```
Are these the same method names one would find in vms?

simon3z replied:
@Fryguy correct.

Fryguy commented:
specs?  I don't see any.  Would be nice to have at least some specs around the code that can be isolated.

Also, great work on this!

simon3z commented:
@Fryguy I addressed your comments and added the specs. Can you review/merge? Thanks!

simon3z commented:
Ping @Fryguy @roliveri @blomquisg @chessbyte review/merge?

Fryguy commented on spec/models/manageiq/providers/kubernetes/container_manager/scanning/job_spec.rb:
```diff
@@ -0,0 +1,28 @@
+
+      MiqServer.stub(:my_server).and_return(
+        FactoryGirl.create(:miq_server, :guid => @guid, :zone => @zone)
+      )
```
There is a pattern for this `@guid, @server, @zone = EvmSpecHelper.create_guid_miq_server_zone`, but I think @kbrock recently cleaned this up even better.

simon3z commented:
> There is a pattern for this @guid, @server, @zone = EvmSpecHelper.create_guid_miq_server_zone, but I think @kbrock recently cleaned this up even better.

@Fryguy I addressed your comment.
